api/features/versioning/serializers.py

```diff
@@ -64,6 +64,163 @@ class Meta:
+        non_model_fields = (
+            "feature_states_to_create",
+            "feature_states_to_update",
+            "segment_ids_to_delete_overrides",
+            "publish_immediately",
+        )
+
+    def create(self, validated_data):
+        for field_name in self.Meta.non_model_fields:
+            validated_data.pop(field_name, None)
+
+        version = super().create(validated_data)
+
+        # Note that we use self.initial_data for handling the feature states here
+        # since we want the raw data (rather than the serialized ORM objects) to
+        # pass into the serializers in the separate private methods used for modifying
+        # the FeatureState objects.
+        for feature_state_to_create in self.initial_data.get(
+            "feature_states_to_create", []
+        ):
+            self._create_feature_state(
+                {**feature_state_to_create, "environment": version.environment_id},
+                version,
+            )
+
+        for feature_state_to_update in self.initial_data.get(
+            "feature_states_to_update", []
+        ):
+            self._update_feature_state(feature_state_to_update, version)
+
+        self._delete_feature_states(
+            self.initial_data.get("segment_ids_to_delete_overrides", []), version
+        )
+
+        if self.validated_data.get("publish_immediately", False):
+            request = self.context["request"]
+            version.publish(
+                published_by=(
+                    request.user if isinstance(request.user, FFAdminUser) else None
+                )
+            )
+
+        return version
+
+    def _create_feature_state(
+        self, feature_state: dict, version: EnvironmentFeatureVersion
+    ) -> None:
+        if not self._is_segment_override(feature_state):
+            raise serializers.ValidationError(
+                {
+                    "feature_states_to_create": "Cannot create FeatureState objects that are not segment overrides."
+                }
+            )
+
+        segment_id = feature_state["feature_segment"]["segment"]
+        if version.feature_states.filter(
+            feature_segment__segment_id=segment_id
+        ).exists():
+            raise serializers.ValidationError(
+                {
+                    "feature_states_to_create": "Segment override already exists for Segment %d"
+                    % segment_id
+                }
+            )
+
+        fs_serializer = EnvironmentFeatureVersionFeatureStateSerializer(
+            data=feature_state,
+            context={
+                "feature": version.feature,
+                "environment": version.environment,
+                "environment_feature_version": version,
+            },
+        )
+        fs_serializer.is_valid(raise_exception=True)
+        fs_serializer.save(
+            environment_feature_version=version,
+            environment=version.environment,
+            feature=version.feature,
+        )
```

Comment:
Is there a DRYer way of doing this? The context was already set in the constructor, it seems weird to have to set them again.

Reply:
I'm not really too sure why we need both the context and the save kwargs. I tried to follow it down the rabbit hole (of my own making), but couldn't see anything obvious. I've tidied it up a little bit to use a single dict definition in both cases, and I'll try to tackle a better refactor in a separate piece of work.